File: hexbound_backend/Hexbound.API/Models/TurnActions.cs
namespace Hexbound.API.Models;

/// <summary>
/// Tracks action economy per actor per turn (D&D 5e rules)
/// </summary>
public class TurnActions
{
    public bool ActionUsed { get; set; } = false;
    public bool BonusActionUsed { get; set; } = false;
    public int MovementRemaining { get; set; } = 6; // Default speed in hexes
    public bool ReactionUsed { get; set; } = false;

    /// <summary>
    /// Reset all actions for a new turn
    /// </summary>
    public void Reset(int speed = 6)
    {
        ActionUsed = false;
        BonusActionUsed = false;
        MovementRemaining = speed;
        ReactionUsed = false;
    }

    /// <summary>
    /// Check if any actions are still available
    /// </summary>
    public bool HasActionsRemaining => !ActionUsed || !BonusActionUsed || MovementRemaining > 0;
}

File: hexbound_client/lib/game/models/turn_actions.dart
/// Turn Actions Model - D&D 5e action economy tracking
class TurnActions {
  bool actionUsed;
  bool bonusActionUsed;
  int movementRemaining;
  bool reactionUsed;

  TurnActions({
    this.actionUsed = false,
    this.bonusActionUsed = false,
    this.movementRemaining = 6,
    this.reactionUsed = false,
  });

  factory TurnActions.fromJson(Map<String, dynamic> json) => TurnActions(
        actionUsed: json['actionUsed'] ?? false,
        bonusActionUsed: json['bonusActionUsed'] ?? false,
        movementRemaining: json['movementRemaining'] ?? 6,
        reactionUsed: json['reactionUsed'] ?? false,
      );

  Map<String, dynamic> toJson() => {
        'actionUsed': actionUsed,
        'bonusActionUsed': bonusActionUsed,
        'movementRemaining': movementRemaining,
        'reactionUsed': reactionUsed,
      };

  bool get hasActionsRemaining =>
      !actionUsed || !bonusActionUsed || movementRemaining > 0;
}
